### ISSUE TITLE
[V3] colorRampTexture with NEAREST textureFilter

### ISSUE THREAD

**Issue by HiroyukiYANO** [mapbox/mapbox-gl-js#12909]
## Motivation

The TextureFilter of colorRampTexture is now only fixed in 'LINEAR' for.
But in some case, the user may wish to use 'NEAREST' for color palette, especially using 'step' expression for raster-color.
No color interpolation is needed near threshold value.

![with-proxy-web weathernews jp_smaphogen1314vs_hack_gis_mapbox_sample_v3_raster-color_radar_v3 0 0-beta 3 html(1280x960) (3)](https://github.com/mapbox/mapbox-gl-js/assets/7903230/198f0387-f059-4595-aa77-122fbfe7d5d3)
![with-proxy-web weathernews jp_smaphogen1314vs_hack_gis_mapbox_sample_v3_raster-color_radar_v3 0 0-beta 3 html(1280x960) (2)](https://github.com/mapbox/mapbox-gl-js/assets/7903230/7d151c31-923f-4a4f-bb6f-1c322880ca47)

## Design
adding new paint property like `raster-color-resampling: "nearest" | "linear"` ("linear" as default)
or use existing paint property `raster-resampling` for raster-color-ramp.

 

### Implementation
change [L130 of draw_raster.js ](https://github.com/mapbox/mapbox-gl-js/blob/c072bf74d3c8c21cbdc73dcae83e04035e07c8c4/src/render/draw_raster.js#L130)

from
```
tex.bind(gl.LINEAR, gl.CLAMP_TO_EDGE);
```
to
```
if(layer.paint.get('raster-color-resampling') === 'nearest') {
    tex.bind(gl.NEAREST, gl.CLAMP_TO_EDGE);
} else {
    tex.bind(gl.LINEAR, gl.CLAMP_TO_EDGE);
}
```

**Comment by mourner** [comment]
cc @rreusser

**Comment by rreusser** [comment]
Good catch, @HiroyukiYANO! I believe what you're seeing is a small misalignment in tabulated color scale texture sampling resulting from the fact that WebGL texture values are defined at the *center* of each pixel. I have submitted a bug fix which will be released in the near future. Until then, you might try the following workaround:

```
"raster-color-range": [ -0.5, 255.5 ],
```

With this code, you should find that raster tile pixel values map precisely to `raster-color` expression values and no longer displays unexpected colors:

```
'raster-color': [
    'step',
    ['raster-value'],
    'transparent', 0, // <- precise integers should work
    '#00DCEA', 2,
    '#0005B9', 8,
    '#FFF000', 16,
    '#F50100'
],
'raster-color-mix': [255, 0, 0, 0],
'raster-color-range': [-0.5, 255.5 ],
```

Once the fix is released, then `"raster-color-range": [0, 255]` should produce the expected results.

You may also find this comment useful, in which I described some of the subtle differences in behavior which result from more of a "categorical" colorization use case instead of a continuous, quantitative color scale: https://github.com/mapbox/mapbox-gl-js/pull/12368#issuecomment-1370120613 Any feedback on the value or shortcomings of this feature are greatly appreciated! 🙇 

Hopefully the above workaround is successful, and if so, I think it's still best to leave this issue open until the fix is released.

**Comment by rreusser** [comment]
Oh, and one small extra note is that you're very correct that a property like `raster-color-resampling` may be appropriate. It may ultimately prove necessary, but it's generally *much much* harder to remove properties than to add them later, so I'd like to explore first whether the above workaround/bugfix remedies the issue. If not, then I'm glad to further explore details and possibilities. Finally, thank you for the clear description of the issue! 🙇

**Comment by HiroyukiYANO** [comment]
Thank you for your kindly reply.
I have tried your workaround.

The result is ... almost OK.

with subtract 0.5 from original step value, then it woks fine.
```
'raster-color': [
    'step',
    ['raster-value'],
    'transparent', 0,
    '#00DCEA', 2 - 0.5, 
    '#0005B9', 8 - 0.5,
    '#FFF000', 16 - 0.5,
    '#F50100'
],
'raster-color-mix': [255, 0, 0, 0],
'raster-color-range': [-0.5, 255.5 ],
```
You can see my code example [here](https://jsfiddle.net/yanoh/skcwtq59/).


<img width="745" alt="Screen Shot 2023-09-27 at 10 58 29" src="https://github.com/mapbox/mapbox-gl-js/assets/7903230/334ddf74-7494-45a6-9811-3c5f95760555">

**Comment by rreusser** [comment]
Ahh, thank you for the followup. I think I failed to shift the expression to compensate for the shifted `raster-color-range` values. Apologies for the continued workarounds and confusion, but if this yields the correct result, I think it's the appropriate solution for now, and when the fix is released, I think you will be able to get rid of the factors of 0.5 throughout.

**Comment by HiroyukiYANO** [comment]
Thank you. I got it.
I'm looking forward to releasing fixed one!
